Implemented Levene's test of homogeneity of variances

charlesknightsbridge opened the pull request:
(for Anova)
also skip Bartlett's test of homogeneity if there is a stratum with a single observation

Results agree with NIST (https://www.itl.nist.gov/div898/handbook/eda/section3/eda35a.htm)

charlesknightsbridge committed:
Implemented Leven's test of homogeneity of variances

(for Anova)
also skip Bartlett's test of homogeneity if there is a stratum with a single observation

torstenchr commented on src/statistics/means.pas:
```diff
@@ -252,29 +255,85 @@ procedure TMeans.DoCalcAnova(ResultDF: TMeansDatafile);
+    end;
+    with AnovaRecord do
+    begin
+      W := (float(obs - k) * z1) / (float(lStrata) * z2);
```
This needs an if statement to ensure that (lStrata * z2) is never 0

torstenchr commented on src/statistics/means.pas:
```diff
@@ -252,29 +255,85 @@ procedure TMeans.DoCalcAnova(ResultDF: TMeansDatafile);
+        inc(idx);
+      end;
+      ytot += ysum;
+      zbar[i] := ysum / N.AsFloat[i];
```
This needs an if statment ensuring that N.AsFloat[i] is never 0

charlesknightsbridge replied:
done

charlesknightsbridge committed:
Check zero denominators in Levene test

Should never happen, but worth checking

torstenchr commented on src/statistics/means.pas:
```diff
@@ -252,29 +255,91 @@ procedure TMeans.DoCalcAnova(ResultDF: TMeansDatafile);
+      if (N.AsInteger[i] > 0) then
+        zbar[i] := ysum / N.AsFloat[i]
+      else
+        FExecutor.Error('Levene test error - n=1 in stratum ' + i.ToString);
```
I'm uncertain, but doesn't this create a memory leak because ResultDF is never freed? The call to .Error(..) is raising an exception and it is not cause until we reach the executor. Hence the ResultDF (and all objects owned by it) will never be freed.

charlesknightsbridge commented:
Right. I need to make this an information message using FOutputGenerator instead. This condition will only be met if there is some coding error, since AGG only provides results with n >= 1. The message is also wrong; should say ‘n = 0’JamieOn May 24, 2025, at 8:48 AM, Torsten Bonde Christiansen ***@***.***> wrote:﻿
@torstenchr requested changes on this pull request.



In src/statistics/means.pas:
> +    setLength(z, Obs);
+    idx := 0;
+    ytot := 0;
+    for i := 0 to lStrata do begin
+      ysum := 0;
+      m := Median.AsFloat[i];
+      for j := 0 to N.AsInteger[i] - 1 do begin
+        z[idx] := abs(CountVar.AsFloat[idx] - m);
+        ysum += z[idx];
+        inc(idx);
+      end;
+      ytot += ysum;
+      if (N.AsInteger[i] > 0) then
+        zbar[i] := ysum / N.AsFloat[i]
+      else
+        FExecutor.Error('Levene test error - n=1 in stratum ' + i.ToString);

I'm uncertain, but doesn't this create a memory leak because ResultDF is never freed? The call to .Error(..) is raising an exception and it is not cause until we reach the executor. Hence the ResultDF (and all objects owned by it) will never be freed.

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you authored the thread.Message ID: ***@***.***>

charlesknightsbridge committed:
Change error action in Levene test

If n=0 in a stratum (programming error), send warning message rather than error
